File: leetcode/2. problem/20.有效的括号.java
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start

class Solution {
    public boolean isValid(String s) {

        // 解法一：利用Stack先进后出的规则特性
        // return extracted(s);

        // 解法二： 解法一的改良
        return extracted2(s);

    }

    /**
     * 方法改良
     * Your runtime beats 98.78 % of java submissions
     * Your memory usage beats 37.75 % of java submissions (39.3 MB)
     * 
     * @param s
     * @return
     */
    private boolean extracted2(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            // 如果是半边的括号就放进去等待匹配
            // 放进对应的右半边括号是为了简化匹配,这样匹配的时候直接判断是不是相等就好了,节省时间
            if (c == '(')
                stack.push(')');
            if (c == '[')
                stack.push(']');
            if (c == '{')
                stack.push('}');
            // 如果是半边的括号就放进去等待匹配

            if (c == ')' || c == ']' || c == '}') {
                // 右括号放进去就开始匹配 一旦匹配失败就失败喽
                // stack.isEmpty()的时候加入右括号,肯定不能自己匹配(左边什么都没有),就失败了
                if (stack.isEmpty() || c != stack.pop())
                    return false;
            }
        }
        return stack.isEmpty();
    }

    /**
     * 
     * Your runtime beats 98.78 % of java submissions
     * Your memory usage beats 18.75 % of java submissions (39.5 MB)
     * 
     * @param s
     * @return
     */
    private boolean extracted(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char tempChar = s.charAt(i);
            // 将所有的括号左边放入到stack中
            if (tempChar == '(' || tempChar == '{' || tempChar == '[') {
                stack.push(tempChar);
            } else {
                // 从stack中抽取元素，对括号进行配对
                if (!stack.isEmpty()) {
                    char stackChar = stack.pop();
                    if (tempChar == ')') {
                        tempChar = '(';
                    } else if (tempChar == '}') {
                        tempChar = '{';
                    } else if (tempChar == ']') {
                        tempChar = '[';
                        // 特性情况直接返回false
                    } else {
                        return false;
                    }
                    // 如果 字符不配对，直接返回false
                    if (tempChar != stackChar) {
                        return false;
                    }
                    // 如果是括号的右侧，那么直接返回false
                } else {
                    return false;
                }
            }
        }
        // 最终判断stack中是否存在未配对的扩号
        return stack.isEmpty();
    }
}
// @lc code=end

File: leetcode/2. problem/golang/20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	// return method1(s)

	return test(s)
}

func test(s string) bool {
	length := len(s)
	if length%2 == 1 {
		return false
	}

	data := map[byte]byte{')': '(', ']': '[', '}': '{'}
	stack := []byte{}
	for i := 0; i < length; i++ {
		if data[s[i]] > 0 {
			if len(stack) == 0 || stack[len(stack)-1] != data[s[i]] {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	return len(stack) == 0
}

func method1(s string) bool {
	length := len(s)
	if length%2 == 1 {
		return false
	}
	// 确认配对方式
	pairs := map[byte]byte{')': '(', '}': '{', ']': '['}
	// 此处定义个一个切片数组
	stack := []byte{}
	for i := 0; i < length; i++ {
		// 如果当前元素在配对中存在
		if pairs[s[i]] > 0 {
			// 如果找不到相互匹配的元素，就退出
			if len(stack) == 0 || stack[len(stack)-1] != pairs[s[i]] {
				return false
			}
			// 移除stack中的元素
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	return len(stack) == 0
}

// @lc code=end

